File: frontend/src/api/collaboration.ts
import axiosClient from "./axiosClient";

export interface AgentMessageItem {
  message_id: string;
  conversation_id: string;
  job_id: string;
  task_id?: string;
  from_agent: string;
  to_agent: string;
  message_type: string;
  payload: Record<string, any>;
  confidence: number;
  status: string;
  timestamp: string;
}

export interface AgentArtifactItem {
  artifact_id: string;
  job_id: string;
  task_id?: string;
  owner_agent: string;
  artifact_type: string;
  title: string;
  metadata: Record<string, any>;
  content: Record<string, any>;
  confidence: number;
  version: number;
  parent_version_id?: string;
  created_at: string;
}

export interface ConsensusDecisionItem {
  consensus_id: string;
  job_id: string;
  task_id?: string;
  topic: string;
  proposals: Array<Record<string, any>>;
  strategy_used: string;
  resolved_output: Record<string, any>;
  winning_agent?: string;
  confidence: number;
  is_conflict: boolean;
  conflict_details?: Record<string, any>;
  resolved_at: string;
}

export interface CollaborationSummary {
  job_id: string;
  delegation_count: number;
  conflict_count: number;
  consensus_count: number;
  message_count: number;
  artifact_count: number;
  active_conversations: Array<Record<string, any>>;
  metrics_summary: Record<string, any>;
}

export interface CollaborationMetrics {
  job_id: string;
  message_count: number;
  artifact_count: number;
  consensus_count: number;
  conflict_count: number;
  delegation_count: number;
  total_sequential_latency_seconds: number;
  actual_job_latency_seconds: number;
  parallel_efficiency: number;
  agent_utilization_percent: Record<string, number>;
}

export interface DelegationPayload {
  from_agent: string;
  target_agent: string;
  task_description: string;
  inputs?: Record<string, any>;
  timeout_seconds?: number;
  max_retries?: number;
  approval_required?: boolean;
}

export const collaborationApi = {
  getMessages: async (jobId: string, params?: { conversation_id?: string; agent_id?: string }) => {
    const res = await axiosClient.get<AgentMessageItem[]>(`/agents/${jobId}/messages`, { params });
    return res.data;
  },

  postMessage: async (jobId: string, payload: { from_agent: string; to_agent: string; message_type?: string; payload?: Record<string, any>; conversation_id?: string }) => {
    const res = await axiosClient.post<AgentMessageItem>(`/agents/${jobId}/message`, payload);
    return res.data;
  },

  getArtifacts: async (jobId: string, params?: { artifact_type?: string; owner_agent?: string }) => {
    const res = await axiosClient.get<AgentArtifactItem[]>(`/agents/${jobId}/artifacts`, { params });
    return res.data;
  },

  getConsensus: async (jobId: string) => {
    const res = await axiosClient.get<ConsensusDecisionItem[]>(`/agents/${jobId}/consensus`);
    return res.data;
  },

  getSummary: async (jobId: string) => {
    const res = await axiosClient.get<CollaborationSummary>(`/agents/${jobId}/collaboration`);
    return res.data;
  },

  getMetrics: async (jobId: string) => {
    const res = await axiosClient.get<CollaborationMetrics>(`/agents/${jobId}/metrics`);
    return res.data;
  },

  delegateTask: async (jobId: string, payload: DelegationPayload) => {
    const res = await axiosClient.post(`/agents/${jobId}/delegate`, payload);
    return res.data;
  },

  connectStream: (jobId: string, onEvent: (event: any) => void) => {
    const token = localStorage.getItem("access_token");
    const url = `/api/v1/agents/${jobId}/stream${token ? `?token=${token}` : ""}`;
    const es = new EventSource(url);
    es.onmessage = (e) => {
      try {
        const data = JSON.parse(e.data);
        onEvent(data);
      } catch {}
    };
    return () => es.close();
  },
};

File: backend/app/agents/schemas/collaboration.py
"""
Pydantic v2 Schemas for Multi-Agent Collaboration Engine REST API.
"""
from typing import List, Dict, Optional, Any
from datetime import datetime
from pydantic import BaseModel, Field, field_validator


class AgentMessageSchema(BaseModel):
    message_id: str
    conversation_id: str
    job_id: str
    task_id: Optional[str] = None
    from_agent: str
    to_agent: str
    message_type: str
    payload: Dict[str, Any] = Field(default_factory=dict)
    confidence: int
    status: str
    timestamp: datetime


class SendMessageRequest(BaseModel):
    from_agent: str = Field(..., description="Sender agent ID")
    to_agent: str = Field(..., description="Recipient agent ID or 'broadcast'/'group'")
    message_type: str = Field("point_to_point", description="point_to_point | broadcast | delegation | proposal")
    payload: Dict[str, Any] = Field(default_factory=dict)
    conversation_id: Optional[str] = Field(None, description="Optional conversation thread ID")
    task_id: Optional[str] = Field(None)


class AgentArtifactSchema(BaseModel):
    artifact_id: str
    job_id: str
    task_id: Optional[str] = None
    owner_agent: str
    artifact_type: str
    title: str
    metadata: Dict[str, Any] = Field(default_factory=dict)
    content: Dict[str, Any] = Field(default_factory=dict)
    confidence: int
    version: int
    parent_version_id: Optional[str] = None
    created_at: datetime


class ConsensusDecisionSchema(BaseModel):
    consensus_id: str
    job_id: str
    task_id: Optional[str] = None
    topic: str
    proposals: List[Dict[str, Any]] = Field(default_factory=list)
    strategy_used: str
    resolved_output: Dict[str, Any] = Field(default_factory=dict)
    winning_agent: Optional[str] = None
    confidence: int
    is_conflict: bool
    conflict_details: Optional[Dict[str, Any]] = None
    resolved_at: datetime


class DelegationRequest(BaseModel):
    from_agent: str = Field(..., description="Delegating agent ID")
    target_agent: str = Field(..., description="Target recipient agent ID")
    task_description: str = Field(..., description="Sub-task goal instructions")
    inputs: Dict[str, Any] = Field(default_factory=dict)
    timeout_seconds: int = Field(30, ge=5, le=300)
    max_retries: int = Field(2, ge=0, le=5)
    approval_required: bool = Field(False)


class DelegationResponse(BaseModel):
    status: str
    confidence: int
    messages: List[str] = Field(default_factory=list)
    outputs: Dict[str, Any] = Field(default_factory=dict)


class CollaborationMetricsSchema(BaseModel):
    job_id: str
    message_count: int
    artifact_count: int
    consensus_count: int
    conflict_count: int
    delegation_count: int
    total_sequential_latency_seconds: float
    actual_job_latency_seconds: float
    parallel_efficiency: float
    agent_utilization_percent: Dict[str, float] = Field(default_factory=dict)


class CollaborationSummaryResponse(BaseModel):
    job_id: str
    delegation_count: int
    conflict_count: int
    consensus_count: int
    message_count: int
    artifact_count: int
    active_conversations: List[Any] = Field(default_factory=list)
    metrics_summary: Dict[str, Any] = Field(default_factory=dict)
